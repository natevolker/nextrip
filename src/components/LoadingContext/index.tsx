import { createContext,useContext, useEffect, useState } from 'react'
import { useRouter } from 'next/router'
import PageLoader from 'next/dist/client/page-loader'
import { getRouteRegex } from 'next/dist/shared/lib/router/utils'
import { RouteRegex } from 'next/dist/shared/lib/router/utils/route-regex'

const getRoute = (url: string, pageList: string[]) => {
  let match: [string, RouteRegex, RegExpExecArray] | undefined
  for (const page of pageList) {
    const routeRegex = getRouteRegex(page)
    const result = routeRegex.re.exec(url)
    if (result) {
      match = [page, routeRegex, result]
      break;
    }
  }

  if (match) {
    const [page, regex, result] = match
    const query: Record<string, string|string[]> = {}
    for (const [key, value] of new URL(url, 'http://a.b').searchParams.entries()) {
      query[key] = page.includes(`[...${key}]`) && !Array.isArray(value)
        ? [value]
        : value
    }
    for (const key in regex.groups) {
      query[key] = page.includes(`[...${key}]`) && !Array.isArray(result[regex.groups[key].pos])
        ? result[regex.groups[key].pos]?.split('/')
        : result[regex.groups[key].pos]
    }
    return {
      pathname: page,
      asPath: url,
      query
    }
  }

  return {
    pathname: '',
    asPath: '',
    query: {}
  }
}

const Context = createContext<{
  pathname: string,
  asPath: string,
  query: Record<string, string | string[]>,
  supress: boolean
}>({} as any)

export const LoadingContext: React.FC = ({ children }) => {
  const router = useRouter()
  const [pageList, setPageList] = useState<string[]>([])
  const [supress, setSupress] = useState(false)
  const [loading, setLoading] = useState({
    pathname: router?.pathname ?? '',
    asPath: router?.asPath ?? '',
    query: (router?.query as Record<string, string | string[]>) ?? {}
  })

  useEffect(() => {
    (async () => {
      setPageList(await new PageLoader('', '').getPageList())
    })()
  }, [])

  useEffect(() => {
    const handleRouteChangeStart = (url: string) => {
      setSupress(true)
      setTimeout(() => { setSupress(false) }, 500)
      const route = getRoute(url, pageList)
      if (route && JSON.stringify(loading) !== JSON.stringify(route)) {
        setLoading(route)
      }
    }

    router?.events.on('routeChangeStart', handleRouteChangeStart)
    return () => {
      router?.events.off('routeChangeStart', handleRouteChangeStart)
    }
  }, [loading, router, pageList])

  return (
    <Context.Provider value={{ ...loading, supress }}>
      {children}
    </Context.Provider>
  )
}

export const useLoading = () => useContext(Context)