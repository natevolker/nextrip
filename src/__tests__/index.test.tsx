import { render, screen } from '@testing-library/preact'
import { LoadingContext } from '@components/LoadingContext'
import Home from '@pages/[[...slug]]'

const fixture = {"stops":[{"stop_id":49441,"latitude":44.975235,"longitude":-93.274242,"description":"9th St S & Nicollet Mall"}],"alerts":[{"stop_closed":false,"alert_text":"Route 9 detoured off Glenwood from Lyndale to 11th St/10th St until further notice due to Green Line Extension(SWLRT) construction"}],"departures":[{"actual":true,"trip_id":"19989416-DEC21-MVS-BUS-Weekday-01","stop_id":49441,"departure_text":"Due","departure_time":1648037780,"description":"Bryn Mawr / West End / Louisiana TC","route_id":"9","route_short_name":"9","direction_id":1,"direction_text":"WB","terminal":"C","schedule_relationship":"Scheduled"},{"actual":false,"trip_id":"19989417-DEC21-MVS-BUS-Weekday-01","stop_id":49441,"departure_text":"7:47","departure_time":1648039620,"description":"Bryn Mawr / West End / Louisiana TC","route_id":"9","route_short_name":"9","direction_id":1,"direction_text":"WB","terminal":"C","schedule_relationship":"Scheduled"},{"actual":false,"trip_id":"19989405-DEC21-MVS-BUS-Weekday-01","stop_id":49441,"departure_text":"8:17","departure_time":1648041420,"description":"Bryn Mawr / West End / Greenbrier","route_id":"9","route_short_name":"9","direction_id":1,"direction_text":"WB","terminal":"N","schedule_relationship":"Scheduled"},{"actual":false,"trip_id":"19989406-DEC21-MVS-BUS-Weekday-01","stop_id":49441,"departure_text":"8:50","departure_time":1648043400,"description":"Bryn Mawr / West End / Greenbrier","route_id":"9","route_short_name":"9","direction_id":1,"direction_text":"WB","terminal":"N","schedule_relationship":"Scheduled"}]}

describe('Home', () => {
  it('renders a heading', () => {
    render(
      <LoadingContext>
        <Home route_id={null} direction_id={null} stop_id={null} routes={null} directions={null} stops={null} departures={null} />
      </LoadingContext>
    )

    const heading = screen.getByText(/Real-Time Departures/)
    expect(heading).toBeInTheDocument()
  })
})

it('renders with no args', () => {
  const { container } = render(
    <LoadingContext>
      <Home route_id={null} direction_id={null} stop_id={null} routes={null} directions={null} stops={null} departures={null} />
    </LoadingContext>
  )
  expect(container).toMatchSnapshot()
})

it('renders with all args', () => {
  const { container } = render(
    <LoadingContext>
      <Home route_id={'1'} direction_id={'1'} stop_id={'1'} routes={[]} directions={[]} stops={[]} departures={fixture} />
    </LoadingContext>
  )
  expect(container).toMatchSnapshot()
})