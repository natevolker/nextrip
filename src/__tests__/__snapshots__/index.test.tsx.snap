// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with all args 1`] = `
<div>
  <form
    class="sidebar sidebar"
  >
    <h2>
      Real-Time Departures
    </h2>
    <label
      class="select  "
      for="route"
      title=""
    >
      <span
        class="label"
      >
        Route
      </span>
      <select
        class="select-element"
        name="route"
        required=""
      >
        <option
          disabled=""
          value=""
        >
          Select a Route
        </option>
      </select>
    </label>
    <label
      class="select disabled "
      for="direction"
      title="You must first select a route"
    >
      <span
        class="label"
      >
        Direction
      </span>
      <select
        class="select-element"
        disabled=""
        name="direction"
        required=""
      >
        <option
          disabled=""
          value=""
        >
          Select a Direction
        </option>
      </select>
    </label>
    <label
      class="select disabled "
      for="stop"
      title="You must first select a route and direction"
    >
      <span
        class="label"
      >
        Stop
      </span>
      <select
        class="select-element"
        disabled=""
        name="stop"
        required=""
      >
        <option
          disabled=""
          value=""
        >
          Select a Stop
        </option>
      </select>
    </label>
  </form>
  <div
    class="primary primary"
  >
    <div
      class="center"
    >
      Select a Route, Direction, and Stop to see departure times.
    </div>
    <div
      class="out"
    />
  </div>
</div>
`;

exports[`renders with no args 1`] = `
<div>
  <form
    class="sidebar sidebar"
  >
    <h2>
      Real-Time Departures
    </h2>
    <label
      class="select  "
      for="route"
      title=""
    >
      <span
        class="label"
      >
        Route
      </span>
      <select
        class="select-element"
        name="route"
        required=""
      >
        <option
          disabled=""
          value=""
        >
          Select a Route
        </option>
      </select>
    </label>
    <label
      class="select disabled "
      for="direction"
      title="You must first select a route"
    >
      <span
        class="label"
      >
        Direction
      </span>
      <select
        class="select-element"
        disabled=""
        name="direction"
        required=""
      >
        <option
          disabled=""
          value=""
        >
          Select a Direction
        </option>
      </select>
    </label>
    <label
      class="select disabled "
      for="stop"
      title="You must first select a route and direction"
    >
      <span
        class="label"
      >
        Stop
      </span>
      <select
        class="select-element"
        disabled=""
        name="stop"
        required=""
      >
        <option
          disabled=""
          value=""
        >
          Select a Stop
        </option>
      </select>
    </label>
  </form>
  <div
    class="primary primary"
  >
    <div
      class="center"
    >
      Select a Route, Direction, and Stop to see departure times.
    </div>
    <div
      class="out"
    />
  </div>
</div>
`;
